nexus/src/sagas.rs

```diff
@@ -475,9 +488,77 @@ async fn sic_create_network_interfaces_undo(
     Ok(())
 }
 
+async fn sic_attach_disks_to_instance(
+    sagactx: ActionContext<SagaInstanceCreate>,
+) -> Result<(), ActionError> {
+    ensure_instance_disk_attach_state(sagactx, true).await
+}
+
+async fn sic_attach_disks_to_instance_undo(
+    sagactx: ActionContext<SagaInstanceCreate>,
+) -> Result<(), anyhow::Error> {
+    Ok(ensure_instance_disk_attach_state(sagactx, false).await?)
+}
+
+async fn ensure_instance_disk_attach_state(
+    sagactx: ActionContext<SagaInstanceCreate>,
+    attached: bool,
+) -> Result<(), ActionError> {
+    let osagactx = sagactx.user_data();
+    let saga_params = sagactx.saga_params();
+    let opctx =
+        OpContext::for_saga_action(&sagactx, &saga_params.serialized_authn);
+
+    let saga_disks = &saga_params.create_params.disks;
+    let instance_name = sagactx.lookup::<db::model::Name>("instance_name")?;
+
+    let organization_name: db::model::Name =
+        saga_params.organization_name.clone().into();
+    let project_name: db::model::Name = saga_params.project_name.clone().into();
+
+    for disk in saga_disks {
+        match disk {
+            params::InstanceDiskAttachment::Create(_) => {
+                todo!();
+            }
+            params::InstanceDiskAttachment::Attach(instance_disk_attach) => {
+                let disk_name: db::model::Name =
+                    instance_disk_attach.disk.clone().into();
+
+                if attached {
+                    osagactx
+                        .nexus()
```

Comment:
Each step in the saga must be idempotent - namely, repeatedly calling `ensure_instance_disk_attach_state`, either as part of the action or undo-action must:

- Result in the same effective state
- Return the same result

Is this currently the case?

(I *think* it is - especially since we're operating at exclusively a DB layer here - but wanted to check that assumption)

Reply:
I'm pretty sure this is true for the same reason but I'm not sure. At the very least, repeated calls to instance_attach_disk and instance_detach_disk is tested by some of the disk related integration tests.

What isn't tested is doing that through the sled agent for a running instance but that code path isn't hit here (?).